src/Orleans.Analyzers/NoRefParamsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,55 @@
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Orleans.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class NoRefParamsDiagnosticAnalyzer : DiagnosticAnalyzer
+    {
+        private const string BaseInterfaceName = "IGrain";
+        public const string DiagnosticId = "ORLEANS0002";
+        public const string Title = "Reference parameter modifiers are not allowed";
+        public const string MessageFormat = Title;
+        public const string Category = "Usage";
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.MethodDeclaration);
+        }
+
+        private static void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            if (!(context.Node is MethodDeclarationSyntax syntax)) return;
+
+            var symbol = context.SemanticModel.GetDeclaredSymbol(syntax);
+            var isClassOrInterface = symbol.ContainingType.TypeKind == TypeKind.Interface ||
+                                     symbol.ContainingType.TypeKind == TypeKind.Class;
+
+            if (!isClassOrInterface) return;
+
+            var implementedInterfaces = symbol.ContainingType
+                                              .AllInterfaces
+                                              .Select(interfaceDef => interfaceDef.Name);
+            if (!implementedInterfaces.Contains(BaseInterfaceName)) return;
+
+            foreach(var param in symbol.Parameters)
+            {
+                if (param.RefKind == RefKind.None) continue;
```

Comment:
Do we want to allow `in` params?

Reply:
I don't know if ```in``` is now supported, if so, they should be allowed.

Reply:
It wouldn't have the same semantics as what people would likely expect, since you can't reference memory in a remote machine, but there's no reason for us to block it.

Let's keep this code in your PR as-is for now and we can always broaden it up later if we decide we want to support it.